### DISCUSSION TITLE
Is there a sane way to inject syntax highlighting into a concatenated string?

### DISCUSSION THREAD

**Discussion by Hubro** [tree-sitter/tree-sitter#1303] [Q&A]
I am implementing a Tree-sitter grammar for YANG, and YANG uses regular expressions heavily. For that reason it would be cool to inject regex syntax highlighting into those strings.

Unfortunately, regexes are often very long, and it's good style in YANG to split them over multiple lines like this:

```yang
    type string {
      pattern '((:|[0-9a-fA-F]{0,4}):)([0-9a-fA-F]{0,4}:){0,5}'
            + '((([0-9a-fA-F]{0,4}:)?(:|[0-9a-fA-F]{0,4}))|'
            + '(((25[0-5]|2[0-4][0-9]|[01]?[0-9]?[0-9])\.){3}'
            + '(25[0-5]|2[0-4][0-9]|[01]?[0-9]?[0-9])))'
            + '(%[\p{N}\p{L}]+)?';
      pattern '(([^:]+:){6}(([^:]+:[^:]+)|(.*\..*)))|'
            + '((([^:]+:)*[^:]+)?::(([^:]+:)*[^:]+)?)'
            + '(%.+)?';
    }
```

If I were to do something like:

```
(statement
  (statement_keyword ["pattern"])
  (argument (string) @regex))
(statement
  (statement_keyword ["pattern"])
  (argument (string_concatenation (string) @regex)))
```

Then each individual string in the concatenation expression would be parsed as a standalone pattern, which would be full of errors.

Is there any way to properly inject syntax highlighting in a case like this? I assume it would require special query syntax to select all the string segments as a single string and pass it to `@regex`.

**Comment by ahelwer** [comment]
If such pattern strings are always identifiable and delimited (like with the keyword `pattern` and a semicolon) you could include `/'[\s|\r?\n]*\+[\s|\r?\n]*'/` as a sort of extra string that can be found anywhere in the pattern. Since you can only define extras globally, maybe you'd have to define a separate grammar for these regexes and do grammar injection? I've yet to use that feature but will in the future.

**Comment by stsewd** [comment]
@Hubro are you using this on nvim-treesitter? Because tree-sitter does support having the same nodes as one document https://tree-sitter.github.io/tree-sitter/using-parsers#multi-language-documents.

For nvim-treesitter, we have the `@combined`capture, but it breaks other things, so isn't ready to use yet.

**Reply by Hubro** [reply]
> Because tree-sitter does support having the same nodes as one document https://tree-sitter.github.io/tree-sitter/using-parsers#multi-language-documents

~~Of course, my question was regarding whether it's possible to highlight the contents of a concatenated string like `"(foo" + "|bar" + "|baz)"` as a single regex like `"(foo|bar|baz)"`.~~ Oh, after checking the link more carefully it seems like that's what you meant, sorry!

I am using nvim-treesitter, yes. Do you have a link for some documentation or a pull request for `@combined`? It sounds interesting but I can't find anything. 🤔

**Reply by theHamsta** [reply]
It's not documented because it isn't working properly yet. You would have to look into the source code (https://github.com/neovim/neovim/blob/3f9a838454e78252cfd90311472978960040d4ec/runtime/lua/vim/treesitter/languagetree.lua#L315, https://github.com/neovim/neovim/blob/3f9a838454e78252cfd90311472978960040d4ec/runtime/lua/vim/treesitter/languagetree.lua#L393-L395) or the official tree-sitter docs to learn the semantics that `@combined` should have for language injection. The idea is that different injections are not parsed as separate documents but glued together to one single document.

**Reply by Hubro** [reply]
Sounds like precisely what I'm looking for 😁 Glad to hear it's on the roadmap
